src/main/cpp/src/cast_string_to_float.cu

```diff
@@ -0,0 +1,730 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cast_string.hpp"
+
+#include <cub/warp/warp_reduce.cuh>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/utilities/integer_utils.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/strings/convert/convert_floats.hpp>
+#include <cudf/utilities/bit.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
+
+using namespace cudf;
+
+namespace spark_rapids_jni {
+
+namespace detail {
+
+__device__ __inline__ bool is_digit(char c) { return c >= '0' && c <= '9'; }
+
+/**
+ * @brief Identify if a character is whitespace.
+ *
+ * @param chr character to test
+ * @return true if character is a whitespace character
+ */
+constexpr bool is_whitespace(char const chr)
+{
+  switch (chr) {
+    case ' ':
+    case '\r':
+    case '\t':
+    case '\n': return true;
+    default: return false;
+  }
+}
+
+template <typename T, size_type block_size>
+class string_to_float {
+ public:
+  __device__ string_to_float(T* _out,
+                             bitmask_type* _validity,
+                             int32_t* _ansi_except,
+                             size_type* _valid_count,
+                             const char* const _chars,
+                             offset_type const* _offsets,
+                             int _warp_id,
+                             uint64_t const* const _ipow,
+                             bitmask_type const* _incoming_null_mask,
+                             size_type const _num_rows)
```

Comment:
I feel something's wrong: This functor takes too many parameters. It would be great if we can decouple them by using multiple functors instead.

Reply:
The design choice was to use a functor to prevent passing so many variables around to each function, but that leaves us passing it all in at the beginning to feed into that object. I don't know how much we could free up as this is a single kernel invocation for speed reasons.

Reply:
Thinking about this more, we can pull the base functor into this function and then call the private functions with parameters, but it feels like too much for this function. If we push that out to a function we're back in the same boat as now as we have to pass all that into that next function.